_posts/2020-02-17-Combining-MongoDB-and-AWS-Lambda.md

```diff
@@ -0,0 +1,310 @@
+* [Cold start vs warm performance](#cold-start-vs-warm-performance)
+* [Performance conclusions](#performance-conclusion)
+* [VPC peering between AWS and MongoDB Atlas](#vpc-peering-connect-your-lambda-functions-with-your-mongodb-atlas-cluster)
+* [Useful links](#useful-links)
+
+
+# MongoDB and AWS Lambda: a successful marriage?
+
+<div style="text-align: center;" >
+  <img src="/img/2020-02-17-Combining-MongoDB-Atlas-and-AWS-Lambda/rings.png" width="15%" height="15%">
+</div>
+
+Can we use MongoDB Atlas when working with AWS Lambda?  
+Yes we can!   
+It's simple enough to setup and above all also performing well!  
+
+This blog concerns mainly two things:
+* Why would you use MongoDB + AWS Lambda and how does it perform
+* How to create a production grade setup with vpc-peering
+
+# MongoDB and AWS Lambda: Why?
+As I am both a fan of AWS Lambda and MongoDB Atlas it was fun for me to marry them.  
+However in the real world we don't do things for fun alone.  
+
+<div style="text-align: center;" >
+  <img src="/img/2020-02-17-Combining-MongoDB-Atlas-and-AWS-Lambda/mongodb-plus-aws-lambda.png" width="70%" height="70%">
+</div>
+
+What are the motives to combine MongoDB Atlas and AWS Lambda?
+
+* In case of MongoDB you provision your cluster and you know what you'll pay for it.
+* MongoDB has a rich language to query within a collection
+* On top of your data in MongoDB you can build rich dashboards for business intelligence.
+* MongoDB supports joins over collections.
+* Support for lot's of indexes
+* Large documents allowed. A MongoDB document can be up to 16 Mb.
+* Flexible data access. Also in a NoSQL world it is important to know your data access model. 
+Data modeling is an important step that needs to be taken carefully before you start persisting your data.
+The way your data is modelled will limit the access patterns that can be used on your data.
+MongoDB offers you more flexibility to access your data in different ways after you have stored it.
+
+# MongoDB and AWS Lambda: Performance
+Suppose you have setup access from your Lambda Functions to your MongoDB Atlas Cluster (If you want to know how, read more about it in the second part of this post).
+
+Since we are dealing with Lambda Functions we also have to deal with cold start. 
+This is the case for any Lambda Function no matter what database it connects to.
+MongoDB mentions on there website that there is an initial startup due to this.
+
+I am testing using the following setup:
+
+<div style="text-align: center;" >
+  <img src="/img/2020-02-17-Combining-MongoDB-Atlas-and-AWS-Lambda/Architecture.png">
+</div>
+
+Items are coming in via requests through the API.
+The `APILambda` drops the items on a queue.  
+A second Lambda Function `SaveMongoDBLambda` stores the items in the MongoDB database.
+
+Other things important to know:
+* In the above setup I limit the maximum number of concurrently running Lambda Functions instances of `SaveMongoDBLambda` to 10.
```

Comment:
Maybe explain why this limit? Don't know if it is important.

Reply:
It's actually to avoid a 1000 cold starts when I would not limit that. I'd rather have some items queueing up then 1000 cold starts of 7 seconds. Since buffering is often why a queue is used.
I have not added more explanation since I think it might create even more confusion for  people unfamiliar with Lambda concurrency.

Reply:
👍 this is not an intro to Lambda's either, fine with me